icicle/include/icicle/program/program.h

```diff
@@ -28,70 +28,93 @@ namespace icicle {
+    void set_as_inputs(std::vector<Symbol<S>>& program_parameters)
     {
-      m_nof_inputs = combine_inputs.size();
-      for (int input_idx = 0; input_idx < m_nof_inputs; input_idx++) {
-        combine_inputs[input_idx].set_as_input(input_idx);
+      for (int parameter_idx = 0; parameter_idx < program_parameters.size(); parameter_idx++) {
+        program_parameters[parameter_idx].set_as_input(parameter_idx);
       }
     }
 
-    // run over the DFG held by result and gemerate the program
-    void generate_program(Symbol<S>& result)
+    // run over the DFG held by program_parameters and generate the program
+    void generate_program(std::vector<Symbol<S>>& program_parameters)
     {
-      m_nof_outputs = 1;
-      result.m_operation->m_variable_idx = m_nof_inputs;
+      // run over the graph and allocate location for all constants
       Operation<S>::reset_visit();
-      allocate_constants(result.m_operation);
+      for (auto& result : program_parameters) {
+        allocate_constants(result.m_operation);
+      }
+
+      // run over the graph and generate the program
       Operation<S>::reset_visit();
-      generate_program(result.m_operation);
+      for (int parameter_idx = 0; parameter_idx < program_parameters.size(); parameter_idx++) {
+        Symbol<S>& cur_symbol = program_parameters[parameter_idx];
+        // the operation not yet calculated
+        if (cur_symbol.m_operation->m_variable_idx == -1) {
+          // set the operation location to the current parameter_idx
+          cur_symbol.m_operation->m_variable_idx = parameter_idx;
+          generate_program(cur_symbol.m_operation);
+          continue;
+        }
+        // the operation already calculated but on a different location
+        if (cur_symbol.m_operation->m_variable_idx != parameter_idx)
+          // copy it to this parameter
+          push_copy_instruction(cur_symbol.m_operation->m_variable_idx, parameter_idx);
+      }
     }
 
     // Program
-    std::vector<InstructionType> m_instructions;
-    std::vector<S> m_constants;
-    int m_nof_inputs = 0;
-    int m_nof_outputs = 0;
+    std::vector<InstructionType> m_instructions; // vector of instructions to execute
+    std::vector<S> m_constants;                  // vector of constants to use
+    int m_nof_parameters = 0;
     int m_nof_constants = 0;
     int m_nof_intermidiates = 0;
 
-    const int get_nof_vars() const { return m_nof_inputs + m_nof_outputs + m_nof_constants + m_nof_intermidiates; }
+    const int get_nof_vars() const { return m_nof_parameters + m_nof_constants + m_nof_intermidiates; }
 
     static inline const int INST_OPCODE = 0;
     static inline const int INST_OPERAND1 = 1;
     static inline const int INST_OPERAND2 = 2;
     static inline const int INST_RESULT = 3;
     inline static int get_opcode(const InstructionType instruction) { return (instruction & 0xFF); }
 
-  private:
+  protected:
+    // default constructor
+    Program() {}
```

Comment:
if you don't need it at all you can delete it with
Program() = delete;

Reply:
I need it for classes that inherit Program